SpotiFire.SpotifyLib/SpotifyLibExtensions.cs

```diff
@@ -45,27 +45,35 @@ public static Search SearchPlaylist(this Session session, string query, int play
 
         private readonly static List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>> waiting = new List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>>();
 
-        private readonly static Timer _timer = new Timer(OnTimerTick, null, Timeout.Infinite, 100);
+        private readonly static Timer _timer = new Timer(20);
+
+        static SessionExtensions()
+        {
+            _timer.Elapsed += (sender, args) => OnTimerTick(null);
+        }
 
         private static void OnTimerTick(object state)
         {
             List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>> finished = null;
+
+            List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>> awaited;
             lock (waiting)
             {
-                if (waiting.Any())
+                // let's make a copy of the list so we don't run into problems removing items while iterating over it
+                awaited = waiting.ToList();
+            }
+            if (awaited.Any())
+            {
+                finished = new List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>>();
+                foreach (Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>> t in awaited)
                 {
-                    finished = new List<Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>>>();
-                    foreach (Tuple<IAsyncLoaded, TaskCompletionSource<IAsyncLoaded>> t in waiting)
+                    if (t.Item1.IsLoaded && t.Item1.IsReady)
                     {
-                        if (t.Item1.IsLoaded && t.Item1.IsReady)
-                        {
-                            finished.Add(t);
-                            waiting.Remove(t);
-                        }
+                        finished.Add(t);
+                        waiting.Remove(t);
```

Comment:
This lacks locking around the List. :)

I guess it would be better to drop the foreach entirely and loop the collection from the rear and use RemoveAt. 

The previous await code has been written by me, so I'm facepalming right now I forgot about the troubles when iterating over a list with foreach and removing items at the same time.